lib/IDE/Refactoring.cpp

```diff
@@ -1463,6 +1463,112 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
+  }
+};
+
+static CollapsibleNestedIfInfo findCollapseNestedIfTarget(ResolvedCursorInfo CursorInfo) {
+  if (CursorInfo.Kind != CursorInfoKind::StmtStart)
+    return CollapsibleNestedIfInfo();
+  struct IfStmtFinder: public SourceEntityWalker {
+    SourceLoc StartLoc;
+    CollapsibleNestedIfInfo IfInfo;
+    IfStmtFinder(SourceLoc StartLoc): StartLoc(StartLoc), IfInfo() {}
+    bool finishedInnerIfButNotFinishedOuterIf() {
+      return IfInfo.InnerIf && !IfInfo.FinishedOuterIf;
+    }
+    bool walkToStmtPre(Stmt *S) {
+      if (finishedInnerIfButNotFinishedOuterIf()) {
+        IfInfo.FoundNonCollapsibleItem = true;
+        return false;
+      }
+
+      bool StmtIsOuterIfBrace =
+        IfInfo.OuterIf && !IfInfo.InnerIf && S->getKind() == StmtKind::Brace;
+      if (StmtIsOuterIfBrace) {
+        return true;
+      }
+
+      auto *IFS = dyn_cast<IfStmt>(S);
+      if (!IFS) {
+        return false;
+      }
+      if (IFS->getCond().size() != 1) {
+        return false;
+      }
+      if (!IfInfo.OuterIf) {
+        IfInfo.OuterIf = IFS;
+        return true;
+      } else {
+        IfInfo.InnerIf = IFS;
+        return false;
+      }
+    }
+    bool walkToStmtPost(Stmt *S) {
+      assert(S != IfInfo.InnerIf && "Should not traverse inner if statement");
+      if (S == IfInfo.OuterIf) {
+        IfInfo.FinishedOuterIf = true;
+      }
+      return true;
+    }
+    bool walkToDeclPre(Decl *D, CharSourceRange Range) {
+      if (finishedInnerIfButNotFinishedOuterIf()) {
+        IfInfo.FoundNonCollapsibleItem = true;
+        return false;
+      }
+      return true;
+    }
+    bool walkToExprPre(Expr *E) {
+      if (finishedInnerIfButNotFinishedOuterIf()) {
+        IfInfo.FoundNonCollapsibleItem = true;
+        return false;
+      }
+      return true;
+    }
+
+  } Walker(CursorInfo.TrailingStmt->getStartLoc());
+  Walker.walk(CursorInfo.TrailingStmt);
+  return Walker.IfInfo;
+}
+
+bool RefactoringActionCollapseNestedIfExpr::isApplicable(ResolvedCursorInfo Tok) {
+    return findCollapseNestedIfTarget(Tok).isValid();
+}
+
+bool RefactoringActionCollapseNestedIfExpr::performChange() {
+    auto Target = findCollapseNestedIfTarget(CursorInfo);
+    if (!Target.isValid())
+        return true;
+    auto OuterIfConditionals = Target.OuterIf->getCond().vec();
+    auto InnerIfConditionals = Target.InnerIf->getCond().vec();
+
+    auto OuterIfConditionalText = Lexer::getCharSourceRangeFromSourceRange(
+     SM, OuterIfConditionals[0].getSourceRange()).str();
+    auto InnerIfConditionalText = Lexer::getCharSourceRangeFromSourceRange(
+     SM, InnerIfConditionals[0].getSourceRange()).str();
+    auto ThenStatementText = Lexer::getCharSourceRangeFromSourceRange(
+     SM, Target.InnerIf->getThenStmt()->getSourceRange()).str();
+
+    llvm::SmallString<64> DeclBuffer;
+    llvm::raw_svector_ostream OS(DeclBuffer);
+    OS << tok::kw_if << " (" << OuterIfConditionalText << ") && (";
```

Comment:
While thinking on how multiple if-statement conditions could be combined together, I have re-discovered once again that in C-based languages comma is an operator and it's computed left-to-right, which means for boolean conditions it's same as `&&`.

So if we have

```swift
if a > 1 {
  if a < 1 || a > 3 {
  }
}
```

instead of `if (a > 1) && (a < 1 || a > 3)` we can just do `if a > 1, a < 1 || a > 3`.
It may look a bit off, but when we have optional binding of pattern matching inside if statements, then `,` works for all cases:

```
if let a = a, a > 1 {
  if let b = b, b < 3 || b > 5 {
    if d == false {
    }
  }
}
```

with `,` combining conditions is very easy:

```swift
if let a = a, a > 1, let b = b, b < 3 || b > 5, d == false {
}
```

Reply:
I was planning on implementing combining of conditionals for `if let` statements using this approach next. If we prefer combining conditionals using commas for all cases, it certainly simplifies the logic. I believe it also makes it so we can easily combine `if` statements with more than one conditional. Thoughts?

Reply:
Actually I just went ahead and did it. Let me know what you think!